### PR TITLE
Major Update to Task

### PR DESCRIPTION
**Pull request by thanrobinson**
- includes spatial density
- includes scenario framework
- adds efficiency to connectivity and vectorization

### CODE REVIEW

**Comment by thanrobinson** [comment]
[Modify ee_rootdir](https://trello.com/c/DA2IwDi7/184-modify-eerootdir)

File: src/task_scl_eff_pot_hab.py [hunk 1]
```diff
@@ -188,49 +165,72 @@ def density_to_patch_size(ft):
...
+                ),
+                1,
+            )
+            .selfMask()
         )
 
-        scl_eff_pot_hab = "{}/geographies/Sumatra/scl_poly/{}/scl_eff_pot_hab".format(
-            self.species, self.taskdate
+        potential_core = (
+            dilate(potential_core, 2)
```
Comment (saanobhaai) [inline src/task_scl_eff_pot_hab.py]: Should the `2` argument be a var so it's clear what it is?

File: src/task_scl_eff_pot_hab.py [hunk 2]
```diff
@@ -188,49 +165,72 @@ def density_to_patch_size(ft):
...
+            .updateMask(scl_polys)
+            .addBands(scl_polys)
+            .addBands(self.probability)
+            .rename(["constant", "scl_poly", "probability"])
+            .reduceToVectors(
+                reducer=ee.Reducer.max(),
+                geometry=self.extent,
+                scale=self.scale,
+                crs="EPSG:4326",
+                maxPixels=1e10,
```
Comment (saanobhaai) [inline src/task_scl_eff_pot_hab.py]: `EETask` has a default
`ee_max_pixels = 500000000000`
which should be used for this arg. You can override it in the `__init__` or at top of class (i.e. with `scale`). Also, elsewhere we've used `1e13` -- any reason not to use the larger number?

File: src/task_scl_eff_pot_hab.py [hunk 3]
```diff
@@ -162,6 +116,29 @@ def density_to_patch_size(ft):
                 None, {"ECO_ID": ft.get("ECO_ID"), "min_core_size": min_core_size}
             )
 
+        def dilate(image, distance):
```
Comment (saanobhaai) [inline src/task_scl_eff_pot_hab.py:122]: This is awesome though a little black-box.

File: src/task_scl_eff_pot_hab.py [hunk 4]
```diff
@@ -162,6 +116,29 @@ def density_to_patch_size(ft):
...
+        current_range = self.historic_range.updateMask(self.extirpated).selfMask()
+        low_hii_mask = self.hii.lte(self.thresholds["hii"]).selfMask()
+        str_hab = (
+            str_hab_mask.updateMask(elev_mask)
+            .updateMask(low_hii_mask)
+            .updateMask(self.water)
+            .updateMask(current_range)
+        )
+
+        connected_potential_habitat = str_hab.connectedPixelCount(625, True).reproject(
```
Comment (saanobhaai) [inline src/task_scl_eff_pot_hab.py]: Should `625` be a var so we know why that many pixels?

File: src/task_scl_eff_pot_hab.py [hunk 5]
```diff
@@ -162,6 +116,29 @@ def density_to_patch_size(ft):
...
+            crs="EPSG:4326", scale=self.scale
+        )
+
+        potential_habitat = connected_potential_habitat.updateMask(
+            connected_potential_habitat.gte(3)
```
Comment (saanobhaai) [inline src/task_scl_eff_pot_hab.py]: `3` should be a variable I think; this could easily change

File: src/task_scl_eff_pot_hab.py [hunk 6]
```diff
@@ -60,10 +55,10 @@ class SCLPolygons(SCLTask):
         },
     }
     thresholds = {
-        "str_lc": 50,
+        "str_lc": 4,
```
Comment (saanobhaai) [inline src/task_scl_eff_pot_hab.py]: I think this var name should be a bit less abbreviated; it wasn't clear to me what it was

**Review by saanobhaai — CHANGES_REQUESTED** [review]
Just a few small things... this is totally great, an example of how more time can lead to fewer lines of more elegant and understandable code.

File: src/task_scl_eff_pot_hab.py (hunk 1 above)
Comment (thanrobinson) [reply src/task_scl_eff_pot_hab.py]: done. now called self.thresholds['connectity_distance']. 
the function takes a value in pixels.  but perhaps the input should be km and we have a function to convert?  Currently running everything at a 1000 km scale is clean in this way, but if we may want to make this flexible if we change scales?

File: src/task_scl_eff_pot_hab.py (hunk 2 above)
Comment (thanrobinson) [reply src/task_scl_eff_pot_hab.py]: good point. done.  I don't think there is a reason not to use the larger number. early on in earth engine developer interactions they made a point to say set this at a reasonable level.  I think it was to make sure people didn't blow up the resources

File: src/task_scl_eff_pot_hab.py (hunk 3 above)
Comment (thanrobinson) [reply src/task_scl_eff_pot_hab.py:122]: a little bit black boxy, but it's nice and fast.  the distance

File: src/task_scl_eff_pot_hab.py (hunk 4 above)
Comment (thanrobinson) [reply src/task_scl_eff_pot_hab.py]: yes, and in fact it already was programmed as a variable

File: src/task_scl_eff_pot_hab.py (hunk 5 above)
Comment (thanrobinson) [reply src/task_scl_eff_pot_hab.py]: same as above

File: src/task_scl_eff_pot_hab.py (hunk 6 above)
Comment (thanrobinson) [reply src/task_scl_eff_pot_hab.py]: renamed to structural_habitat

**Pull request merged**